wsmux/stream.go

```diff
@@ -0,0 +1,326 @@
+package wsmux
+
+import (
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// DefaultCapacity is the maximum length the read buffer will accept
+const DefaultCapacity = 256
+
+type stream struct {
+	id      uint32
+	bufLock sync.Mutex // lock b of type []byte
```

Comment:
Try to start over with one lock and one condition.
You have following variables: 
```go
type stream struct{
  m sync.Mutex
  c sync.Cond
  buffer []byte // fixed size circular buffer
  bufStart int
  bufEnd int
  endErr error // set when stream breaks or ends, set io.EOF if at end of stream
  unblocked int // number of bytes remote is willing to receive
}
```
From Read() you return endErr, if buffer is bufStart == bufEnd, and endErr is not nil... Otherwise you wait for cond..

On write you wait until unblocked > zero... Then send unblocked number of bytes...

When ack arrives you add n to unblocked and signal cond...

If more data arrives than fits in the circular buffer... Remote side has violated the ack messages.. and we close the steam with an error..

Okay, I could be wrong... But this is just an outline...

To make thing simple, I suggest we make a method on stream to handle each frame... So all session does is handle syn frames and forward all other frames to appropriate stream.

This is just an idea, but I think you have too many synchronization primitives here... It makes it harder and bugs more likely... You need to define the invariants for you code... In the outline I gave above the invariants are:
You must hold m to modify any property, and c must be signaled with broadcast after any change.

This is simple. When waiting for a specific condition to happen, like endErr is set, just do:
```go
s.m.Lock()
defer s.m.Unlock()

for s.endErr != nil {
  s.c.Wait()
}

// Now my condition is the case and I hold the lock
```
This way you can do really complicated conditions...

Note: this is traditionally​ how cond and mutex are used in c/c++, not sure you've ever used those before?
In go channels are great, but when things gets stateful cond/mutex can be better... And much faster...